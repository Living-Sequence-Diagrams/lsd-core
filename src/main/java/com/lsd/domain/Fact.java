package com.lsd.domain;

import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class Fact {
    String key;
    String value;
}
