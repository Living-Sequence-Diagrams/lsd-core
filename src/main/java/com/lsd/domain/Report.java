package com.lsd.domain;

import lombok.Builder;
import lombok.Value;

import java.util.Collection;

@Value
@Builder
public class Report {
    String title;
    Collection<Scenario> scenarios;
}
