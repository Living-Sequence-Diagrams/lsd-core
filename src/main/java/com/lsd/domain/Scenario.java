package com.lsd.domain;

import com.lsd.PlantUmlRenderer;
import com.lsd.SvgConverter;
import com.lsd.domain.interactions.Interaction;
import lombok.Builder;
import lombok.Value;

import java.util.Collection;

@Value
@Builder
public class Scenario {
    String title;
    String description;
    Collection<Interaction> interactions;
    Collection<Fact> facts;

    PlantUmlRenderer plantUmlRenderer = new PlantUmlRenderer();
    SvgConverter svgConverter = new SvgConverter();

    public String sequenceDiagram() {
        String plantUml = plantUmlRenderer.render(interactions);
        return svgConverter.convert(plantUml);
    }
}
