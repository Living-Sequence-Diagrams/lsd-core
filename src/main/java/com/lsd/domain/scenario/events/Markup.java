package com.lsd.domain.scenario.events;

import lombok.RequiredArgsConstructor;
import lombok.Value;

@Value
@RequiredArgsConstructor
public class Markup implements Event {
    String value;

    public String toMarkup() {
        return value;
    }
}
