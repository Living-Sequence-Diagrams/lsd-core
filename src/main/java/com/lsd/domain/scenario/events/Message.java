package com.lsd.domain.scenario.events;

import lombok.Builder;
import lombok.Value;

import static java.lang.String.format;

@Value
@Builder
public class Message implements Interaction {
    String id;
    String label;
    String from;
    String to;
    Object data;

    public String toMarkup() {
        return format("%s -> %s:[[#%s %s]]",
                getFrom(), getTo(), getId(), getLabel());
    }
}
