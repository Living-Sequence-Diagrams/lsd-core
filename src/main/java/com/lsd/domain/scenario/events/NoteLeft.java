package com.lsd.domain.scenario.events;

import lombok.RequiredArgsConstructor;
import lombok.Value;

import static java.lang.String.format;

@Value
@RequiredArgsConstructor
public class NoteLeft implements Event {
    String value;

    public String toMarkup() {
        return format("note left: %s", getValue());
    }
}
