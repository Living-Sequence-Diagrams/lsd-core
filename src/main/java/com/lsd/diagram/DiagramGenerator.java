package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
