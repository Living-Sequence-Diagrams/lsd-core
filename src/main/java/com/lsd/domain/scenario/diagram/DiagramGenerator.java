package com.lsd.domain.scenario.diagram;

import com.lsd.domain.Participant;
import com.lsd.domain.scenario.events.Event;
import com.lsd.domain.scenario.events.Interaction;
import com.lsd.properties.LsdProperties;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<Event> events;

    public SequenceDiagram sequenceDiagram() {
        return SequenceDiagram.builder()
                .uml(generateSequenceUml())
                .interactions(events.stream()
                        .filter(Interaction.class::isInstance)
                        .map(Interaction.class::cast)
                        .collect(toList()))
                .build();
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(Event::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
