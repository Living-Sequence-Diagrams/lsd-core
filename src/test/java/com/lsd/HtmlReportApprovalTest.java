package com.lsd;

import com.lsd.domain.scenario.Fact;
import com.lsd.domain.Report;
import com.lsd.domain.scenario.Scenario;
import com.lsd.domain.scenario.diagram.DiagramGenerator;
import com.lsd.domain.scenario.events.Message;
import com.lsd.domain.scenario.events.NoteLeft;
import com.lsd.domain.scenario.events.Markup;
import com.lsd.domain.scenario.events.ResponseMessage;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static com.lsd.domain.ParticipantType.*;

class HtmlReportApprovalTest {
    private final HtmlReportRenderer htmlReportRenderer = new HtmlReportRenderer();

    @Test
    void renderHtml() {
        String report = htmlReportRenderer.render(Report.builder()
                .title("LSD")
                .scenarios(List.of(
                        Scenario.builder()
                                .title("First Scenario")
                                .description("<pre>Given a Cat\nAnd a dog\nWhen the dog barks\nThen the cat hisses</pre>")
                                .facts(List.of(
                                        Fact.builder()
                                                .key("Greeting")
                                                .value("Hi")
                                                .build())
                                )
                                .sequenceDiagram(DiagramGenerator.builder()
                                        .includes(Set.of())
                                        .participants(List.of(
                                                ACTOR.called("A", "Arnie"),
                                                BOUNDARY.called("Unused participant"),
                                                BOUNDARY.called("Another Unused participant", "G"),
                                                DATABASE.called("B", "Badboy \\nDB")
                                        ))
                                        .events(List.of(
                                                Message.builder()
                                                        .id("111")
                                                        .label("Sending a greeting")
                                                        .from("A")
                                                        .to("B")
                                                        .data("Hi")
                                                        .build(),
                                                new Markup("..."),
                                                ResponseMessage.builder()
                                                        .id("222")
                                                        .label("Sending a response")
                                                        .from("B")
                                                        .to("A")
                                                        .data("Yo")
                                                        .build(),
                                                new NoteLeft("Informal")
                                        ))
                                        .build()
                                        .sequenceDiagram())
                                .build(),
                        Scenario.builder()
                                .title("Second Scenario")
                                .description("<p><ul><li>Some details</li><li>Some more details</li></ul></p>")
                                .sequenceDiagram(DiagramGenerator.builder()
                                        .includes(Set.of())
                                        .participants(List.of())
                                        .events(List.of(
                                                Message.builder()
                                                        .id("333")
                                                        .label("An interaction")
                                                        .from("Alpha")
                                                        .to("Beta")
                                                        .data("{\"name\": \"alpha\", \"description\":\"Something long to see how the popup window will respond to text that is wider than the initial box size\"}")
                                                        .build(),
                                                Message.builder()
                                                        .id("444")
                                                        .label("An interaction")
                                                        .from("Beta")
                                                        .to("Gamma")
                                                        .data("β")
                                                        .build(),
                                                Message.builder()
                                                        .id("555")
                                                        .label("An interaction")
                                                        .from("Gamma")
                                                        .to("Delta")
                                                        .data("γ")
                                                        .build()))
                                        .build().sequenceDiagram())
                                .build()))
                .build());

        Approvals.verifyHtml(report);
    }
}