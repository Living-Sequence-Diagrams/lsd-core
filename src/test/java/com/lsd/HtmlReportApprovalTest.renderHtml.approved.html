<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <title>LSD</title>
    <style>
        .overlay {
            position: fixed;
            top: 0;
            bottom: 0;
            left: 0;
            right: 0;
            background: rgba(0, 0, 0, 0.7);
            transition: opacity 500ms;
            visibility: hidden;
            opacity: 0;
        }

        .overlay:target {
            visibility: visible;
            opacity: 1;
        }

        .popup {
            margin: 70px auto;
            padding: 20px;
            background: #fff;
            border-radius: 5px;
            max-width: 70%;
            position: relative;
            transition: all 5s ease-in-out;
        }

        h1, h2, h3 {
            margin-top: 0;
            color: #333;
            font-family: Tahoma, Arial, sans-serif;
        }

        .popup .close {
            position: absolute;
            top: 20px;
            right: 30px;
            transition: all 200ms;
            font-size: 30px;
            font-weight: bold;
            text-decoration: none;
            color: #333;
        }

        .popup .close:hover {
            color: #06D85F;
        }

        .popup .content {
            overflow: auto;
        }

        /*    Make Facts look like table*/
        .factTable {
            display: table;
        }

        .factRow {
            display: table-row;
        }

        .factRowHead {
            display: table-header-group;
            font-weight: bold;
        }

        .factCell, .factRowHead {
            display: table-cell;
            padding: 3px 10px;
            border: 1px solid #999999;
        }

        /* Some basic spacing for now  */
        section {
            border: 15px solid skyblue;
            padding: 50px;
            margin: 20px;
        }
    </style>
</head>
<body>
<main role="main">
    <h1>LSD</h1>
        <article>
        <h2>First Scenario</h2>
        <section class="description">
            <h3>Description</h3>
            <pre>Given a Cat
And a dog
When the dog barks
Then the cat hisses</pre>        </section>
                <section class="facts">
            <h3>Key Facts</h3>
            <div class="factTable">
                <div class="factRow">
                    <div class="factRowHead">Key</div>
                    <div class="factRowHead">Value</div>
                </div>
                                <div class="factRow">
                    <div class="factCell">Greeting</div>
                    <div class="factCell">Hi</div>
                </div>
                            </div>
        </section>
                <section class="sequence diagram svg">
            <h3>Sequence Diagram</h3>
            <?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="297px" preserveAspectRatio="none" style="width:280px;height:297px;" version="1.1" viewBox="0 0 280 297" width="280px" zoomAndPan="magnify">
  <defs>
    <filter height="300%" id="f1ag2z6790r5m8" width="300%" x="-1" y="-1">
      <feGaussianBlur result="blurOut" stdDeviation="2.0" />
      <feColorMatrix in="blurOut" result="blurOut2" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0" />
      <feOffset dx="4.0" dy="4.0" in="blurOut2" result="blurOut3" />
      <feBlend in="SourceGraphic" in2="blurOut3" mode="normal" />
    </filter>
  </defs>
  <g>
    <line style="stroke:#A80036;stroke-width:1.0;" x1="103" x2="103" y1="88.4883" y2="131.6883" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="103" x2="103" y1="131.6883" y2="159.6883" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="103" x2="103" y1="159.6883" y2="212.8883" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="249.5" x2="249.5" y1="88.4883" y2="131.6883" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="249.5" x2="249.5" y1="131.6883" y2="159.6883" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="249.5" x2="249.5" y1="159.6883" y2="212.8883" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="37" x="82" y="85.5352">Arnie</text>
    <ellipse cx="103.5" cy="15" fill="#FEFECE" filter="url(#f1ag2z6790r5m8)" rx="8" ry="8" style="stroke:#A80036;stroke-width:2.0;" />
    <path d="M103.5,23 L103.5,50 M90.5,31 L116.5,31 M103.5,50 L90.5,65 M103.5,50 L116.5,65 " fill="none" filter="url(#f1ag2z6790r5m8)" style="stroke:#A80036;stroke-width:2.0;" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="37" x="82" y="225.4234">Arnie</text>
    <ellipse cx="103.5" cy="238.3766" fill="#FEFECE" filter="url(#f1ag2z6790r5m8)" rx="8" ry="8" style="stroke:#A80036;stroke-width:2.0;" />
    <path d="M103.5,246.3766 L103.5,273.3766 M90.5,254.3766 L116.5,254.3766 M103.5,273.3766 L90.5,288.3766 M103.5,273.3766 L116.5,288.3766 " fill="none" filter="url(#f1ag2z6790r5m8)" style="stroke:#A80036;stroke-width:2.0;" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="50" x="219.5" y="69.0469">Badboy</text>
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="18" x="237.5" y="85.5352">DB</text>
    <path d="M231.5,19.5117 C231.5,9.5117 249.5,9.5117 249.5,9.5117 C249.5,9.5117 267.5,9.5117 267.5,19.5117 L267.5,45.5117 C267.5,55.5117 249.5,55.5117 249.5,55.5117 C249.5,55.5117 231.5,55.5117 231.5,45.5117 L231.5,19.5117 " fill="#FEFECE" filter="url(#f1ag2z6790r5m8)" style="stroke:#000000;stroke-width:1.5;" />
    <path d="M231.5,19.5117 C231.5,29.5117 249.5,29.5117 249.5,29.5117 C249.5,29.5117 267.5,29.5117 267.5,19.5117 " fill="none" style="stroke:#000000;stroke-width:1.5;" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="50" x="219.5" y="225.4234">Badboy</text>
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="18" x="237.5" y="241.9117">DB</text>
    <path d="M231.5,254.8648 C231.5,244.8648 249.5,244.8648 249.5,244.8648 C249.5,244.8648 267.5,244.8648 267.5,254.8648 L267.5,280.8648 C267.5,290.8648 249.5,290.8648 249.5,290.8648 C249.5,290.8648 231.5,290.8648 231.5,280.8648 L231.5,254.8648 " fill="#FEFECE" filter="url(#f1ag2z6790r5m8)" style="stroke:#000000;stroke-width:1.5;" />
    <path d="M231.5,254.8648 C231.5,264.8648 249.5,264.8648 249.5,264.8648 C249.5,264.8648 267.5,264.8648 267.5,254.8648 " fill="none" style="stroke:#000000;stroke-width:1.5;" />
    <polygon fill="#A80036" points="237.5,119.6883,247.5,123.6883,237.5,127.6883,241.5,123.6883" style="stroke:#A80036;stroke-width:1.0;" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="103.5" x2="243.5" y1="123.6883" y2="123.6883" />
    <a href="#111" target="_top" title="#111" xlink:actuate="onRequest" xlink:href="#111" xlink:show="new" xlink:title="#111" xlink:type="simple">
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="83" x="110.5" y="115.0566">Sending food</text>
    </a>
    <a href="#111" target="_top" title="#111" xlink:actuate="onRequest" xlink:href="#111" xlink:show="new" xlink:title="#111" xlink:type="simple">
      <image height="19" width="19" x="197.5" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABMAAAATCAIAAAD9MqGbAAAA9ElEQVR4Xp2TMQqDMBSGU0S6dquCnQrewsnByYMIbt6idCgUurk7uXiH3qHgIoJ0KCiCYKXwmlSD6SuE1J9v0OT/4hsigaUheEFI3+MVMdh8PMDzYLuF1QoIYRgGWylLVBTMuobdbmqPOM7Xq22zc7F5u4GmsW1RjuP5ebNhBV2HPBfMppk1sS1yOMB+Pz13HTejCPfk+D43kwTvyTkeuUlzvUKWKUGbY5h5v8P5zE5S4XKBquLm7zxy1mt4vRaZ9JK07cek01sWmKYqacqnHQYoij94PrkZBHgeOa7LzdMJ78kJQ27Sr6O7LoE2x+C/TD3LzTeit8koBlhA2AAAAABJRU5ErkJggg==" y="102.4883" />
    </a>
    <line style="stroke:#A80036;stroke-width:1.0;" x1="103.5" x2="113.5" y1="187.9436" y2="183.9436" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="103.5" x2="113.5" y1="187.9436" y2="191.9436" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:2.0,2.0;" x1="103.5" x2="248.5" y1="187.9436" y2="187.9436" />
    <a href="#222" target="_top" title="#222" xlink:actuate="onRequest" xlink:href="#222" xlink:show="new" xlink:title="#222" xlink:type="simple">
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="122" x="120.5" y="183.2014">Sending a response</text>
    </a>
    <path d="M5,164.6883 L5,193.6883 L95,193.6883 L95,174.6883 L85,164.6883 L5,164.6883 " fill="#FBFB77" filter="url(#f1ag2z6790r5m8)" style="stroke:#A80036;stroke-width:1.0;" />
    <path d="M85,164.6883 L85,174.6883 L95,174.6883 L85,164.6883 " fill="#FBFB77" style="stroke:#A80036;stroke-width:1.0;" />
    <text fill="#000000" font-family="sans-serif" font-size="13" lengthAdjust="spacing" textLength="46" x="11" y="186.1461">Friends</text>
    <image height="19" width="19" x="61" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABMAAAATCAIAAAD9MqGbAAABCUlEQVR4XmP4/bucPMSAxv/+vRRTERB9+VKCJoLQ6eyswMLCxMDAwMfHnppqABdPTtYHigDF2dmZ3dwUf/woQ+j8/LlYQICDARWIiHB+/FgkKcmDJg4EEM0gnf7+auiSYMDMDHICJnB1VYTqRJchBIAm/vpFlk4geP48j0ydX7+WgHTKyvKhy+AFGhrCUH/ev5+FKzCwAqB6qE4gOnkyHl0eB9i7NxIRnxC0a1ckIyMjukJUsHy5P1w9SurbsCEEXS0SWLs2CFkxerpdvNgPXQcYTJnihqYSXScQzZjhgaZt+nQPTGVYdAJRX58LXFt1tTWmApw6gai52Q6obeJEV0wpAjoJIvJ1AgCBF8EtYqXtzAAAAABJRU5ErkJggg==" y="169.6883" />
  </g>
</svg>

        </section>
        <section class="sequence diagram uml">
            <h3>Sequence Diagram Markup</h3>
            <pre>@startuml

hide unlinked

  !include &lt;tupadr3/font-awesome-5/heart&gt;
  !include &lt;tupadr3/font-awesome-5/hamburger&gt;

  actor A as &quot;Arnie&quot;
  boundary UnusedParticipant
  boundary AnotherUnusedParticipant as &quot;G&quot;
  database B as &quot;Badboy \nDB&quot;

  A -&gt; B:[[#111 Sending food &lt;$hamburger{scale=0.4}&gt;]]
  ...
  B --&gt;&gt; A:[[#222 Sending a response]]
  note left: Friends &lt;$heart{scale=0.4}&gt;

@enduml</pre>
        </section>
                <!-- Section containing data for sequence diagram popups -->
        <div id="111" class="overlay">
            <div class="popup">
                <h2>Sending food </h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>Hi</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="222" class="overlay">
            <div class="popup">
                <h2>Sending a response</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>Thank You!</pre>
                </div>
            </div>
        </div>
            </article>
        <article>
        <h2>Second Scenario</h2>
        <section class="description">
            <h3>Description</h3>
            <p><ul><li>Some details</li><li>Some more details</li></ul></p>        </section>
                <section class="sequence diagram svg">
            <h3>Sequence Diagram</h3>
            <?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="191px" preserveAspectRatio="none" style="width:409px;height:191px;" version="1.1" viewBox="0 0 409 191" width="409px" zoomAndPan="magnify">
  <defs>
    <filter height="300%" id="f145zwv19czbk0" width="300%" x="-1" y="-1">
      <feGaussianBlur result="blurOut" stdDeviation="2.0" />
      <feColorMatrix in="blurOut" result="blurOut2" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 .4 0" />
      <feOffset dx="4.0" dy="4.0" in="blurOut2" result="blurOut3" />
      <feBlend in="SourceGraphic" in2="blurOut3" mode="normal" />
    </filter>
  </defs>
  <g>
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="34" x2="34" y1="40.4883" y2="148.4199" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="146.5" x2="146.5" y1="40.4883" y2="148.4199" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="260" x2="260" y1="40.4883" y2="148.4199" />
    <line style="stroke:#A80036;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="372.5" x2="372.5" y1="40.4883" y2="148.4199" />
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="54" x="5" y="5" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="40" x="12" y="25.5352">Alpha</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="54" x="5" y="147.4199" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="40" x="12" y="167.9551">Alpha</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="43" x="123.5" y="5" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="29" x="130.5" y="25.5352">Beta</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="43" x="123.5" y="147.4199" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="29" x="130.5" y="167.9551">Beta</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="66" x="225" y="5" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="52" x="232" y="25.5352">Gamma</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="66" x="225" y="147.4199" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="52" x="232" y="167.9551">Gamma</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="49" x="346.5" y="5" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="35" x="353.5" y="25.5352">Delta</text>
    <rect fill="#FEFECE" filter="url(#f145zwv19czbk0)" height="30.4883" style="stroke:#A80036;stroke-width:1.5;" width="49" x="346.5" y="147.4199" />
    <text fill="#000000" font-family="sans-serif" font-size="14" lengthAdjust="spacing" textLength="35" x="353.5" y="167.9551">Delta</text>
    <polygon fill="#A80036" points="135,67.7988,145,71.7988,135,75.7988,139,71.7988" style="stroke:#A80036;stroke-width:1.0;" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="34" x2="141" y1="71.7988" y2="71.7988" />
    <a href="#333" target="_top" title="#333" xlink:actuate="onRequest" xlink:href="#333" xlink:show="new" xlink:title="#333" xlink:type="simple">
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="89" x="41" y="67.0566">An interaction</text>
    </a>
    <polygon fill="#A80036" points="248,97.1094,258,101.1094,248,105.1094,252,101.1094" style="stroke:#A80036;stroke-width:1.0;" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="147" x2="254" y1="101.1094" y2="101.1094" />
    <a href="#444" target="_top" title="#444" xlink:actuate="onRequest" xlink:href="#444" xlink:show="new" xlink:title="#444" xlink:type="simple">
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="89" x="154" y="96.3672">An interaction</text>
    </a>
    <polygon fill="#A80036" points="361,126.4199,371,130.4199,361,134.4199,365,130.4199" style="stroke:#A80036;stroke-width:1.0;" />
    <line style="stroke:#A80036;stroke-width:1.0;" x1="260" x2="367" y1="130.4199" y2="130.4199" />
    <a href="#555" target="_top" title="#555" xlink:actuate="onRequest" xlink:href="#555" xlink:show="new" xlink:title="#555" xlink:type="simple">
      <text fill="#0000FF" font-family="sans-serif" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="89" x="267" y="125.6777">An interaction</text>
    </a>
  </g>
</svg>

        </section>
        <section class="sequence diagram uml">
            <h3>Sequence Diagram Markup</h3>
            <pre>@startuml

hide unlinked



  Alpha -&gt; Beta:[[#333 An interaction]]
  Beta -&gt; Gamma:[[#444 An interaction]]
  Gamma -&gt; Delta:[[#555 An interaction]]

@enduml</pre>
        </section>
                <!-- Section containing data for sequence diagram popups -->
        <div id="333" class="overlay">
            <div class="popup">
                <h2>An interaction</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>{"name": "alpha", "description":"Something long to see how the popup window will respond to text that is wider than the initial box size"}</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="444" class="overlay">
            <div class="popup">
                <h2>An interaction</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>β</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="555" class="overlay">
            <div class="popup">
                <h2>An interaction</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>γ</pre>
                </div>
            </div>
        </div>
            </article>
    </main>
</body>
</html>
